galley/tools/gen-meta/main.go

```diff
@@ -0,0 +1,247 @@
+//  Copyright 2018 Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"html/template"
+	"io/ioutil"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/ghodss/yaml"
+)
+
+const usage = `
+
+gen-meta [runtime|kube] <input-yaml-path> <output-go-path> 
+
+`
+
+var knownProtoTypes = map[string]struct{}{
+	"google.protobuf.Struct": {},
+}
+
+// metadata is a combination of read and derived metadata.
+type metadata struct {
+	Items           []*entry   `json:"items"`
+	ProtoGoPackages []string   `json:"-"`
+	ProtoDefs       []protoDef `json:"-"`
+}
+
+// entry in a metadata file
+type entry struct {
+	Kind      string `json:"kind"`
+	ListKind  string `json:"listKind"`
+	Singular  string `json:"singular"`
+	Plural    string `json:"plural"`
+	Group     string `json:"group"`
+	Version   string `json:"version"`
+	Proto     string `json:"proto"`
+	Gogo      bool   `json:"gogo"`
+	Converter string `json:"converter"`
+}
+
+// proto related metadata
+type protoDef struct {
+	MessageName string `json:"-"`
+	Gogo        bool   `json:"-"`
+}
+
+func main() {
+	if len(os.Args) != 5 {
+		fmt.Printf(usage)
+		fmt.Printf("%v\n", os.Args)
+		os.Exit(-1)
+	}
+
+	// The tool can generate both K8s level, as well as the proto level metadata.
+	isRuntime := false
```

Comment:
Any other possible option here besides runtime vs. kube? A boolean flag might be better.